### ISSUE TITLE
Changing Primary Key of non-existing Object

### ISSUE THREAD

**Issue by LordNali** [realm/realm-cocoa#4543]
## Goals
I'd like to replace an Object with a Primary Key (String) in a Realm. Therefore I remove the existing Object (in fact I remove ALL Objects from the Realm) and add the same Object with the new Primary Key.

## Expected Results
After deleting the existing Object, the new Object should be added to the Realm.

## Actual Results
Adding the Object with the new Primary Key throws an Exception.

`Terminating app due to uncaught exception 'RLMException', reason: 'Can't create object with existing primary key value '05472F21-6C2E-412B-9144-2725463DEBE3'.'`

## Steps to Reproduce
See Code Sample.

## Code Sample

```swift
// Get existing Object from Realm
guard let realmObject = realm.objects(MyRealmObject.self).first else {
 // no object found in realm
 return
}

// Convert Realm Object to an internal Class
// Remark: I have two Classes - MyObject and MyRealmObject. 
// Both can be initialized with the other. I know this seems weird, 
// but actually necessary in our case.
// The internal class only uses values types of the realm object, 
// so there´s no 'link' between them
let internalObject = MyObject(realmObject: realmObject)

// Change Id of the internal Object
// which becomes / is the Primary Key of the corresponding Realm Object
internalObject.id = newId // (A String)
            
// Convert internal Object to Realm Object
let newRealmObject = MyRealmObject(internalObject: internalObject)

// Delete all Objects from Realm and add the new Object
try realm.write {
 realm.deleteAll()
 realm.add(newRealmObject, update: false)
}
```

In a seconds attempt, I split the two realm statements (deleteAll and add) into two separate realm.write requests - but still the same exception:

```swift
try realm.write {
 realm.deleteAll()
}

try realm.write {
 realm.add(newRealmObject, update: false)
}
```

Now the weird part, when I set the 'update' Argument in realm.add to **true**, the statement works. So my guess would be, there´s still an object in the realm. But how so? After I used the statements deleteAll()? And furthermore: I tried to retrieve any existing Objects in the Realm before adding the new one, the result of this query has always count 0 - so there should be no Object inside the Realm!? But why do I get then this exception?
 
## Version of Realm and Tooling
Realm version: 2.1.1
Xcode version: 8.1
iOS/OSX version: 10.1 / Sierra
Dependency manager + version: Carthage

**Comment by stel** [comment]
Hi @LordNali! Could you please share an example Xcode project to reproduce this case?

`Realm.deleteAll()` deletes all objects from the realm, so you should be able to create a new objects with "old" primary keys, please check out this example:

```Swift
class TestObject: Object {
    dynamic var id: String = ""

    override static func primaryKey() -> String? {
        return "id"
    }
}

...

let realm = try! Realm()

let firstObject = TestObject()
firstObject.id = "1"

try! realm.write {
    realm.add(firstObject)
}

let secondObject = TestObject()
secondObject.id = "1"

try! realm.write {
    realm.deleteAll()
    realm.add(secondObject)
}
```

**Comment by LordNali** [comment]
@stel Sharing the Project is unfortunately not an option: It´s huge and the´re many customer specific extensions we´re not allowed to publish.

But maybe you could help me in another way: I´m pretty sure that that `realm.deleteAll()` works perfectly (like in your example). So my guess: Is there any possibility that my `internalObject`(which is initialized using the Object from the Realm) holds some kind of reference to the Realm Object? And by that calling `realm.deleteAll()` removes the object from the Realm, but it´s still persist in memory due to to pointer from the internal Object?

`realm.objects(MyRealmObject).count` is `0`after executing `realm.deleteAll()`, but maybe the Realm Object still exists in Memory due to a (unknown) pointer from the internal Object? My internal Object uses only value types from the realm object - so there shouldn't be any link / pointer to the realm object after the Initialization, but maybe I´m missing something here?

**Comment by stel** [comment]
@LordNali unfortunately without any reproduction case or code examples it's difficult guess what's going on there. I suggest you to create a simple example and move step by step from there. Also make sure that you create a new object only once and there are no other threads that do the same.

**Comment by istx25** [comment]
@LordNali: Any luck with @stel's request?

**Comment by LordNali** [comment]
I'm still trying to find the cause of the problem - a simple / simplified example works fine - but recreating the circumstances from the original situation is quiet difficult.

**Comment by stel** [comment]
Setting the [Exception breakpoint in Xcode](http://stackoverflow.com/questions/17802662/exception-breakpoint-in-xcode) can be helpful in finding the exact line in your code where this issue happens.

**Comment by LordNali** [comment]
Finally - I found the cause: Inside my Realm Object (`Parent`) I'm referencing other Realm Objects (`Children`). In my first Code-Examaple I deleted the Parent Object, but forgot to remove the Children Objects. I should know this, because I posted a Issue regarding this Topic some Time ago. And furthermore #1186 .

Now in my second Code Sample I explicitly used `realm.deleteAll()` to remove all Objects from the Realm. This should have in fact solve the cascading delete Issue - but unfortunately there´s was another problem in my Code: A race condition with another write request for the same Object from a different Thread:

1. Thread A calls `realm.deleteAll()`
2. Thread B calls `realm.deleteAll()`
3. Thread A/B calls `realm.write`
4. Thread B/A calls `realm.write` - and there the crash happens, an Object with the same primary key already exists

So in the end there we're two causes for the error
1. `realm.delete(object)` is not enough, if there are linked Objects inside the Realm Object
2. Ensure simultaneous `realm.delete()` and `realm.write()` calls from different Thread on the same Realm-File are executed **without** interruption

@stel Using Exception breakpoint in Xcode was new to me and really helpful, thanks a lot!
